experiments/process_sandbox/library_runner.cc

```diff
@@ -239,184 +219,531 @@ namespace sandbox
-    for (size_t i = 0; i < size_bits - SUPERSLAB_BITS; i++)
+
+    // fprintf(stderr, "Child starting\n");
+    // printf(
+    //"Child trying to map fd %d at addr %p (0x%zx)\n", SharedMemRegion, addr,
+    // length);
+    void* ptr = mmap(
+      addr,
+      length,
+      PROT_READ | PROT_WRITE,
+      MAP_FIXED_NOREPLACE | MAP_SHARED | platform::detail::map_nocore,
+      SharedMemRegion,
+      0);
+
+    // printf("%p\n", ptr);
+    if (ptr == MAP_FAILED)
+    {
+      err(1, "Mapping shared heap failed");
+    }
+
+    shared = reinterpret_cast<SharedMemoryRegion*>(ptr);
+    // Splice the pagemap page inherited from the parent into the pagemap.
+    void* pagemap_chunk = GlobalPagemap::pagemap().page_for_address(
+      reinterpret_cast<uintptr_t>(ptr));
+    munmap(pagemap_chunk, 4096);
+    void* shared_pagemap = mmap(
+      pagemap_chunk, 4096, PROT_READ, MAP_SHARED | MAP_FIXED, PageMapPage, 0);
+    if (shared_pagemap == MAP_FAILED)
     {
-      size_t run = 1ULL << i;
-      GlobalPagemap::pagemap().set_range(
-        ss, (uint8_t)(64 + i + SUPERSLAB_BITS), run);
-      ss = ss + SUPERSLAB_SIZE * run;
+      err(1, "Mapping shared pagemap page failed");
     }
+    shared_memory_start = shared->start;
+    shared_memory_end = shared->end;
+    SANDBOX_INVARIANT(
+      shared_pagemap == pagemap_chunk,
+      "Mapping pagmemap chunk failed.  Expected {}, got {}",
+      pagemap_chunk,
+      shared_pagemap);
+    (void)shared_pagemap;
+
+    done_bootstrapping = true;
   }
 
-  void ProxyPageMap::clear_large_size(void* p, size_t size)
+  using sandbox::platform::Handle;
+  using Socket = sandbox::platform::SocketPair::Socket;
+
+  /**
+   * The socket that is used for upcalls to the parent process.
+   */
+  Socket upcallSocket;
+
+  /**
+   * Perform an upcall.  This takes the kind of upcall, the data to be sent,
+   * and the file descriptor to send as arguments.  The file descriptor may be
+   * -1, in which case the it is not sent.
+   *
+   *  The return value is the integer result of the upcall and a `Handle` that
+   *  is either invalid or the returned file descriptor.
+   *
+   *  This function should not be called directly, it should be invoked via the
+   *  wrapper.
+   */
+  std::pair<uintptr_t, Handle>
+  upcall(sandbox::UpcallKind k, void* buffer, size_t size, int fd)
+  {
+    Handle out_fd(fd);
+    UpcallRequest req{k, size, reinterpret_cast<uintptr_t>(buffer)};
+    upcallSocket.send(&req, sizeof(req), out_fd);
+    out_fd.take();
+    int depth = ++shared->token.upcall_depth;
+    (void)depth;
+    shared->token.is_child_executing = false;
+    shared->token.parent.wake();
+    runloop(depth);
+    Handle in_fd;
+    UpcallResponse response;
+    upcallSocket.receive(&response, sizeof(response), in_fd);
+    return {response.response, std::move(in_fd)};
+  }
+
+  /**
+   * Perform an upcall, of the specified kind, passing `data`.  The `data`
+   * argument must point to the shared heap.
```

Comment:
Shouldn't this function check the pointer is inside the shared heap? Instead of relying on the callers to do that? This is a publicly visible function, no?